File: apps/server/src/server/routes/files.ts
// /api/files — workspace file access for the UI. Every client-supplied path goes through
// `deps.safePath`, which realpaths the roots and the target before anything touches the disk.

import { promises as fs } from "node:fs";
import path from "node:path";
import type { Hono } from "hono";
import type { FileEntry } from "@perch/protocol";
import { readJsonBody, str } from "../http/body.js";
import { conflict } from "../http/errors.js";
import {
  assertNotStale,
  entryFor,
  exists,
  listDir,
  readTextFile,
  writeSqlFile,
} from "../services/workspace-files.js";
import { safeFileName } from "../services/workspace-paths.js";
import type { RouteDeps } from "../create-server.js";

export function registerFilesRoutes(app: Hono, deps: RouteDeps): void {
  const { safePath, watcher, workspaceRoots } = deps;

  app.get("/api/files", async (c) => {
    const dir = c.req.query("dir");
    if (!dir) {
      // No dir given: hand back the workspace roots themselves.
      const entries: FileEntry[] = [];
      for (const root of await workspaceRoots()) {
        try {
          entries.push(await entryFor(root));
        } catch {
          /* a workspace that no longer exists */
        }
      }
      return c.json(entries);
    }
    return c.json(await listDir(await safePath(dir, "dir")));
  });

  app.get("/api/files/content", async (c) => {
    const full = await safePath(c.req.query("path"));
    const { content, modifiedAt } = await readTextFile(full);
    return c.json({ path: full, content, modifiedAt });
  });

  app.put("/api/files/content", async (c) => {
    const body = await readJsonBody(c);
    const full = await safePath(str(body.path));
    const content = typeof body.content === "string" ? body.content : "";
    const ifModifiedAt = str(body.ifModifiedAt);
    if (ifModifiedAt) {
      const stale = await assertNotStale(full, ifModifiedAt);
      if (stale) throw conflict("the file changed on disk since it was read", stale, "stale_write");
    }
    const modifiedAt = await writeSqlFile(full, content);
    // Our own write must not come back to the client as an external change.
    watcher?.expectWrite(full, modifiedAt);
    return c.json({ path: full, modifiedAt });
  });

  app.post("/api/files", async (c) => {
    const body = await readJsonBody(c);
    const dir = await safePath(str(body.dir), "dir");
    const name = safeFileName(String(body.name ?? ""));
    const full = await safePath(path.join(dir, name));
    if (await exists(full)) throw conflict(`${name} already exists`);
    const modifiedAt = await writeSqlFile(full, "");
    watcher?.expectWrite(full, modifiedAt);
    return c.json(await entryFor(full), 201);
  });

  app.post("/api/files/rename", async (c) => {
    const body = await readJsonBody(c);
    const full = await safePath(str(body.path));
    const name = safeFileName(String(body.name ?? ""));
    const target = await safePath(path.join(path.dirname(full), name));
    if (target !== full && (await exists(target))) throw conflict(`${name} already exists`);
    const { content } = await readTextFile(full);
    if (target !== full) watcher?.expectDelete(full);
    const modifiedAt = await writeSqlFile(target, content);
    watcher?.expectWrite(target, modifiedAt);
    if (target !== full) await fs.rm(full, { force: true });
    return c.json(await entryFor(target));
  });

  app.delete("/api/files", async (c) => {
    const full = await safePath(c.req.query("path"));
    watcher?.expectDelete(full);
    await fs.rm(full, { force: true });
    return c.json({ ok: true });
  });
}

File: apps/server/server/files.go
package server

import (
	"net/http"
	"os"
	"path/filepath"

	"perch/httpx"
	"perch/protocol"
)

// Every client-supplied path goes through SafePath, which realpaths the roots and the target
// before anything touches the disk.
func (s *Server) registerFiles() {
	s.handle("GET /api/files", func(w http.ResponseWriter, r *http.Request) error {
		dir := r.URL.Query().Get("dir")
		if dir == "" {
			// No dir given: hand back the workspace roots themselves.
			roots, err := s.WorkspaceRoots()
			if err != nil {
				return err
			}
			entries := make([]protocol.FileEntry, 0, len(roots))
			for _, root := range roots {
				if entry, err := entryFor(root); err == nil {
					entries = append(entries, entry)
				}
			}
			return httpx.JSON(w, entries)
		}
		full, err := s.SafePath(dir, "dir")
		if err != nil {
			return err
		}
		entries, err := listDir(full)
		if err != nil {
			return err
		}
		return httpx.JSON(w, entries)
	})

	s.handle("GET /api/files/content", func(w http.ResponseWriter, r *http.Request) error {
		full, err := s.SafePath(r.URL.Query().Get("path"), "file")
		if err != nil {
			return err
		}
		content, modifiedAt, err := readTextFile(full)
		if err != nil {
			return err
		}
		return httpx.JSON(w, map[string]any{"path": full, "content": content, "modifiedAt": modifiedAt})
	})

	s.handle("PUT /api/files/content", func(w http.ResponseWriter, r *http.Request) error {
		var body struct {
			Path         string `json:"path"`
			Content      string `json:"content"`
			IfModifiedAt string `json:"ifModifiedAt"`
		}
		if err := httpx.DecodeJSON(r, &body); err != nil {
			return err
		}
		full, err := s.SafePath(body.Path, "file")
		if err != nil {
			return err
		}
		if body.IfModifiedAt != "" {
			if stale := staleWrite(full, body.IfModifiedAt); stale != nil {
				return httpx.Conflict("the file changed on disk since it was read", stale, "stale_write")
			}
		}
		modifiedAt, err := writeSQLFile(full, body.Content)
		if err != nil {
			return err
		}
		// Our own write must not come back to the client as an external change.
		s.expectWrite(full, modifiedAt)
		return httpx.JSON(w, map[string]any{"path": full, "modifiedAt": modifiedAt})
	})

	s.handle("POST /api/files", func(w http.ResponseWriter, r *http.Request) error {
		var body struct {
			Dir  string `json:"dir"`
			Name string `json:"name"`
		}
		if err := httpx.DecodeJSON(r, &body); err != nil {
			return err
		}
		dir, err := s.SafePath(body.Dir, "dir")
		if err != nil {
			return err
		}
		name, err := SafeFileName(body.Name)
		if err != nil {
			return err
		}
		full, err := s.SafePath(filepath.Join(dir, name), "file")
		if err != nil {
			return err
		}
		if fileExists(full) {
			return httpx.Conflict(name+" already exists", nil)
		}
		modifiedAt, err := writeSQLFile(full, "")
		if err != nil {
			return err
		}
		s.expectWrite(full, modifiedAt)
		entry, err := entryFor(full)
		if err != nil {
			return err
		}
		return httpx.JSON(w, entry, http.StatusCreated)
	})

	s.handle("POST /api/files/rename", func(w http.ResponseWriter, r *http.Request) error {
		var body struct {
			Path string `json:"path"`
			Name string `json:"name"`
		}
		if err := httpx.DecodeJSON(r, &body); err != nil {
			return err
		}
		full, err := s.SafePath(body.Path, "file")
		if err != nil {
			return err
		}
		name, err := SafeFileName(body.Name)
		if err != nil {
			return err
		}
		target, err := s.SafePath(filepath.Join(filepath.Dir(full), name), "file")
		if err != nil {
			return err
		}
		if target != full && fileExists(target) {
			return httpx.Conflict(name+" already exists", nil)
		}
		content, _, err := readTextFile(full)
		if err != nil {
			return err
		}
		if target != full {
			s.expectDelete(full)
		}
		modifiedAt, err := writeSQLFile(target, content)
		if err != nil {
			return err
		}
		s.expectWrite(target, modifiedAt)
		if target != full {
			os.Remove(full)
		}
		entry, err := entryFor(target)
		if err != nil {
			return err
		}
		return httpx.JSON(w, entry)
	})

	s.handle("DELETE /api/files", func(w http.ResponseWriter, r *http.Request) error {
		full, err := s.SafePath(r.URL.Query().Get("path"), "file")
		if err != nil {
			return err
		}
		s.expectDelete(full)
		os.Remove(full)
		return httpx.JSON(w, map[string]any{"ok": true})
	})
}

func (s *Server) expectWrite(path, modifiedAt string) {
	if s.watcher != nil {
		s.watcher.ExpectWrite(path, modifiedAt)
	}
}

func (s *Server) expectDelete(path string) {
	if s.watcher != nil {
		s.watcher.ExpectDelete(path)
	}
}
